### PR TITLE
Trashbin endpoint in DAV

### LINKED ISSUE #1332: API for the trash-bin
To implement the Trash Bin into the iOS app we need an API for the trash-bin.v
Requested here: https://help.nextcloud.com/t/wishlist-trash/3049/2

cc @schiessle @croessner @marinofaggiana

_Copied over from https://github.com/nextcloud/nextcloud.com/issues/128_

### PR DESCRIPTION
**Pull request by rullzer**
Fixes #1332

- [x] Add a new DAV collection
- [x] List all files in the trashbin for this user
- [x] Deleting files from trashbin
- [x] Get files from trashbin (just read)
- [x] Empty trashbin (just a delete on it)
- [x] Restore from trashbin 
   * To special restore node that will maitain current behavior (with conflict handling etc).
- [x] Add dav properties
  - [x] Original name
  - [x] Original location

Future work:
-  Restore from trashbin
  * Manual move to files
-  Migrate web UI to DAV
-  Add integration tests

### CODE REVIEW

**Comment by rullzer** [comment]
Ready for review I would say.

Basic functionality is there.

File: apps/dav/lib/RootCollection.php [hunk 1]
```diff
@@ -55,7 +55,7 @@ public function __construct() {
 		);
 		$groupPrincipalBackend = new GroupPrincipalBackend($groupManager);
 		// as soon as debug mode is enabled we allow listing of principals
-		$disableListing = !$config->getSystemValue('debug', false);
+		$disableListing = false; //!$config->getSystemValue('debug', false);
```
Comment (MorrisJobke) [inline apps/dav/lib/RootCollection.php]: Is this wanted?

Comment (rullzer) [reply apps/dav/lib/RootCollection.php]: Ah no... I'll fix that :S

Comment (rullzer) [reply apps/dav/lib/RootCollection.php]: fixed

File: apps/files_trashbin/lib/Sabre/TrashFolder.php [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	public function createDirectory($name) {
+		throw new Forbidden();
+	}
+
+	public function getChild($name) {
+		$entries = \OCA\Files_Trashbin\Helper::getTrashFiles($this->getName(), $this->userId);
+
+		foreach ($entries as $entry) {
+			if ($entry->getName() === $name) {
+				if ($entry->getMimetype() === 'httpd/unix-directory') {
```
Comment (icewind1991) [inline apps/files_trashbin/lib/Sabre/TrashFolder.php]: `$entry->isDirectory()`

File: apps/files_trashbin/lib/Sabre/TrashFolder.php [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+				return new TrashFolderFolder($this->getName(), $this->userId, $entry, $this->getOriginalLocation());
+			}
+			return new TrashFolderFile($this->getName(), $this->userId, $entry, $this->getOriginalLocation());
+		}, $entries);
+
+		return $children;
+	}
+
+	public function childExists($name) {
+		$entries = \OCA\Files_Trashbin\Helper::getTrashFiles($this->getName(), $this->userId);
```
Comment (icewind1991) [inline apps/files_trashbin/lib/Sabre/TrashFolder.php]: this seems horribly inefficient

**Review by MorrisJobke — APPROVED** [review]
Tested and works :+1:

**Review by icewind1991 — DISMISSED** [review]
Also I would prefer to have the trash listing in `trashbin/$user/trash` so the restore folder doesn't have to be in the trash listing

**Comment by MorrisJobke** [comment]
CI failure unrelated

File: apps/files_trashbin/lib/Sabre/TrashFolder.php (hunk 3 above)
Comment (rullzer) [reply apps/files_trashbin/lib/Sabre/TrashFolder.php]: It is. But good enough for now

File: apps/files_trashbin/lib/Sabre/TrashFolder.php (hunk 2 above)
Comment (rullzer) [reply apps/files_trashbin/lib/Sabre/TrashFolder.php]: Well that doesn't exist on https://github.com/nextcloud/server/blob/master/lib/public/Files/FileInfo.php ;)

Comment (rullzer) [reply apps/files_trashbin/lib/Sabre/TrashFolder.php]: But I can use getType

File: apps/files_trashbin/lib/Sabre/TrashFolder.php (hunk 3 above)
Comment (rullzer) [reply apps/files_trashbin/lib/Sabre/TrashFolder.php]: Basically this code is now doing pretty much the same as the original. Baby steps we can optimize later.

**Comment by rullzer** [comment]
@icewind1991 `trashbin/$user/trash` looks weird to me... But I get the point... @MorrisJobke opinions?

File: apps/files_trashbin/lib/Sabre/TrashFolder.php (hunk 2 above)
Comment (icewind1991) [reply apps/files_trashbin/lib/Sabre/TrashFolder.php]: Confused it with `IFileInfo` from the smb lib again :see_no_evil:

**Comment by MorrisJobke** [comment]
> @icewind1991 trashbin/$user/trash looks weird to me... But I get the point... @MorrisJobke opinions?

Makes sense 👍

**Comment by rullzer** [comment]
OK let me fix that later this evening

**Comment by rullzer** [comment]
@icewind1991 please have another look

**Comment by rullzer** [comment]
@icewind1991 ping

**Review by icewind1991 — APPROVED** [review]

**Pull request merged**